### PR TITLE
Gtk4 port

### PR DESCRIPTION
**Pull request by jwahlstrand**
This is ready to try now.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -27,8 +27,8 @@ Colors = "0.9, 0.10, 0.11, 0.12"
 FileIO = "1.6"
 FlameGraphs = "0.2.10, 1"
 Graphics = "0.4, 1"
-Gtk = "1.2"
-GtkObservables = "1"
+Gtk4 = "0.5"
```
Comment (nhz2) [inline Project.toml:30]:
```suggestion
Gtk4 = "0.6"
```

File: src/ProfileView.jl [hunk 2]
```diff
@@ -233,67 +234,64 @@ function viewgui(fcolor, gdict::NestedGraphDict; data=nothing, lidict=nothing, w
...
+            tb_zoom_in = GtkButton(:icon_name,"zoom-in-symbolic")
+            Gtk4.tooltip_text(tb_zoom_in, "zoom in")
+            tb_zoom_out = GtkButton(:icon_name, "zoom-out-symbolic")
+            Gtk4.tooltip_text(tb_zoom_out, "zoom out")
+            tb_info = GtkButton(:icon_name, "dialog-information-symbolic")
+            Gtk4.tooltip_text(tb_info, "ProfileView tips")
+            tb_text = GtkEntry()
+            Gtk4.has_frame(tb_text, false)
+            Gtk4.sensitive(tb_text, false)
+            push!(tb, tb_text)
```
Comment (nhz2) [inline src/ProfileView.jl:274]:
```suggestion
            tb_text.hexpand = true
```
Setting this prevented the text from getting cutoff.

File: src/ProfileView.jl [hunk 3]
```diff
@@ -233,67 +234,64 @@ function viewgui(fcolor, gdict::NestedGraphDict; data=nothing, lidict=nothing, w
...
+            push!(tb, GtkSeparator(:h))
             push!(tb, tb_zoom_fit)
             push!(tb, tb_zoom_out)
             push!(tb, tb_zoom_in)
-            push!(tb, SeparatorToolItem())
+            push!(tb, GtkSeparator(:h))
             push!(tb, tb_info)
-            push!(tb, SeparatorToolItem())
-            push!(tb, tb_text_item)
+            push!(tb, GtkSeparator(:h))
```
Comment (nhz2) [inline src/ProfileView.jl:284]:
```suggestion
            push!(tb, GtkSeparator(:h))
            push!(tb, tb_text)
```
Moving `push!(tb, tb_text)` preserved the order of the toolbar.

File: src/ProfileView.jl [hunk 4]
```diff
@@ -484,12 +481,13 @@ end
...
     c, data, lidict, g = profdata
-    selection = save_dialog("Save profile data as *.jlprof file", toplevel(c), ("*.jlprof",))
-    isempty(selection) && return nothing
-    if data === nothing && lidict === nothing
-        return _save(selection, g)
+    save_dialog("Save profile data as *.jlprof file", toplevel(c), ("*.jlprof",)) do selection
+        isempty(selection) && return nothing
+        if data === nothing && lidict === nothing
+            _save(selection, g)
+        end
```
Comment (nhz2) [inline src/ProfileView.jl:488]:
```suggestion
        if data === nothing && lidict === nothing
            _save(selection, g)
        else
            _save(selection, data, lidict)
        end
```
This fixes the saving functionality.

**Review by nhz2 — CHANGES_REQUESTED** [review]
Looks good with a few patches. I tested this on Windows using the default example. Here is a screenshot:
image

File: Project.toml (hunk 1 above)
Comment (jwahlstrand) [reply Project.toml]: I'd like to keep 0.5 compat as well.

**Review by nhz2 — APPROVED** [review]
Looks good. I also tested on Linux using Pop!_OS 22.04.
Everything in the GUI works like in the previous version.

I get the warning:
```
(process:31294): Gtk-WARNING **: 17:38:16.184: No IM module matching GTK_IM_MODULE=ibus found
```
when opening the viewer, and the following when saving or opening a "testdata.jlprof" file.
```
(process:31294): Gtk-WARNING **: 17:46:52.667: Attempting to add 'file:///home/nathan/github/ProfileView.jl/testdata.jlprof' to the list of recently used resources, but no name of the application that is registering it was defined
```
But these don't seem to cause any issues.

image

**Comment by jwahlstrand** [comment]
I don't see those warnings on Fedora, but they're basically harmless. I wish GTK had a mechanism to silence stuff like that.

**Comment by IanButterworth** [comment]
What's the status of this?

On MacOS it seems to work (though the text on the canvas is a little blurry?) and I see no terminal warnings.

Screenshot 2024-08-04 at 10 04 31 PM

**Comment by IanButterworth** [comment]
I wonder if the low resolution issue I'm seeing is related to the failure of this test on MacOS.. perhaps the range of sz is too small
https://github.com/timholy/ProfileView.jl/blob/f3754d0d50344f0d708e20e61a4d0588eac59582/test/runtests.jl#L151-L159

**Comment by jwahlstrand** [comment]
> I get the warning:
> 
> ```
> (process:31294): Gtk-WARNING **: 17:38:16.184: No IM module matching GTK_IM_MODULE=ibus found
> ```
> 
> when opening the viewer, and the following when saving or opening a "testdata.jlprof" file.
> 
> ```
> (process:31294): Gtk-WARNING **: 17:46:52.667: Attempting to add 'file:///home/nathan/github/ProfileView.jl/testdata.jlprof' to the list of recently used resources, but no name of the application that is registering it was defined
> ```

The second of these warnings should be gone in the most recent version of Gtk4.

**Comment by jwahlstrand** [comment]
On the resolution issue, I wonder if the scaling should be modified for HiDPI displays? That's what GLMakie does.

**Comment by IanButterworth** [comment]
Seems like the blurry text rendering is a known and unsolved issue https://gitlab.gnome.org/GNOME/gtk/-/issues/3787

If you think this is ready, I say we go ahead.

**Comment by timholy** [comment]
Sorry for my inattention (too many updates needed for Julia 1.11 in the Revise/Cthulhu/SnoopCompile etc stack).

I'll start playing with this, but @jwahlstrand feel free to merge whenever you are ready (you have an invite now).

**Comment by jwahlstrand** [comment]
Thanks, I'll check one more thing this evening. The weird thing about the font issue is, we're not using GTK to draw the text, it's all Cairo. For me, it's still quite readable, so I think it's worth going ahead.

I'll try to update Yggdrasil to GTK 4.14 soon.

**Comment by jwahlstrand** [comment]
The CI error came back and I think it was happening because the Profile didn't contain much of anything on MacOS, for some reason. Increasing the number of iterations in the profiled function seems to have fixed it.

I'm seeing an issue locally, only on my Mac. When ProfileView precompiles, using it right away doesn't work. Sometimes the windows don't appear, sometimes I see the following error, no matter what I try to profile:
```
Warning: There were no samples collected.
│ Run your program longer (perhaps by running it multiple times),
│ or adjust the delay between samples with `Profile.init()`.
```

But if I quit Julia and restart it, ProfileView works fine. So it seems like precompiling is having some sort of side effect. It's puzzling because ImageView has similar precompile code and works fine. I want to look into this a bit more before merging.

**Comment by timholy** [comment]
Thanks for your careful work here! I don't have a Mac, nor any good ideas except to wonder if Mac isn't using COW across forked processes? https://softwareengineering.stackexchange.com/questions/244664/global-variable-in-a-linux-shared-library

**Comment by IanButterworth** [comment]
I think there's a bigger issue on macOS. See https://github.com/JuliaLang/julia/issues/53816#issuecomment-2283798709

**Comment by jwahlstrand** [comment]
I commented out the precompile code that causes issues on my Mac (which involves observable handlers). I don't notice any difference in performance so I'm going to merge this PR and investigate later.

**Comment by timholy** [comment]
Thank you @jwahlstrand ! You're a true hero!

**Pull request merged**